cumulusci/tasks/metadata_etl/base.py

```diff
@@ -0,0 +1,293 @@
+        pass
+
+    def _run_task(self):
+        with tempfile.TemporaryDirectory() as tempdir:
+            self._create_directories(tempdir)
+            if self.retrieve:
+                self._retrieve()
+            self._transform()
+            if self.deploy:
+                result = self._deploy()
+                self._post_deploy(result)
+
+
+class BaseMetadataSynthesisTask(BaseMetadataETLTask, metaclass=ABCMeta):
+    """Base class for Metadata ETL tasks that generate new metadata
+    and deploy it into the org, but do not retrieve."""
+
+    deploy = True
+
+    def _generate_package_xml(self, deploy):
+        """Synthesize a package.xml for generated metadata."""
+        generator = PackageXmlGenerator(str(self.deploy_dir), self.api_version)
+        return generator()
+
+    def _transform(self):
+        self._synthesize()
+
+    @abstractmethod
+    def _synthesize(self):
+        """Create new metadata in self.deploy_dir."""
+        pass
+
+
+class BaseMetadataTransformTask(BaseMetadataETLTask, metaclass=ABCMeta):
+    """Base class for Metadata ETL tasks that extract metadata,
+    transform it, and deploy it back into the org."""
+
+    retrieve = True
+    deploy = True
+
+    @abstractmethod
+    def _get_entities(self):
+        """Return a dict of Metadata API entities and API names to be transformed."""
+        pass
+
+    def _get_types_package_xml(self):
+        """Generate package.xml content based on the return value of _get_entities()."""
+        base = """    <types>
+{members}
+        <name>{name}</name>
+    </types>
+"""
+        types = ""
+        for entity, api_names in self._get_entities().items():
+            members = "\n".join(
+                f"        <members>{api_name}</members>" for api_name in api_names
+            )
+            types += base.format(members=members, name=entity)
+
+        return types
+
+    def _get_package_xml_content(self, operation):
+        return f"""<?xml version="1.0" encoding="UTF-8"?>
+<Package xmlns="http://soap.sforce.com/2006/04/metadata">
+{self._get_types_package_xml()}
+    <version>{self.api_version}</version>
+</Package>
+"""
+
+    @abstractmethod
+    def _transform(self):
+        pass
+
+
+class MetadataSingleEntityTransformTask(BaseMetadataTransformTask, metaclass=ABCMeta):
+    """Base class for a Metadata ETL task that affects one or more
+    instances of a specific metadata entity. Concrete subclasses must set
+    `entity` to the Metadata API entity transformed, and implement _transform_entity()."""
+
+    entity = None
+
+    task_options = {
+        "api_names": {"description": "List of API names of entities to affect"},
+        **BaseMetadataETLTask.task_options,
+    }
+
+    def _init_options(self, kwargs):
+        super()._init_options(kwargs)
+
+        self.api_names = list(
+            map(
+                self._inject_namespace,
+                process_list_arg(self.options.get("api_names", [])),
+            )
+        )
```

Comment:
I just noticed this one, so I won't block approval on it, but this could be:

    self.api_names = [self._inject_namespace(name) for name in process_list_arg(self.options.get("api_names", []))]